staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go

```diff
@@ -162,26 +163,36 @@ func newController() (*Controller, *fakecloud.Cloud, *fake.Clientset) {
 	recorder := broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "service-controller"})
 
 	controller := &Controller{
-		cloud:            cloud,
-		kubeClient:       kubeClient,
-		clusterName:      "test-cluster",
-		cache:            &serviceCache{serviceMap: make(map[string]*cachedService)},
-		eventBroadcaster: broadcaster,
-		eventRecorder:    recorder,
-		nodeLister:       newFakeNodeLister(nil),
-		nodeListerSynced: nodeInformer.Informer().HasSynced,
-		serviceQueue:     workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "service"),
-		nodeQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "node"),
-		lastSyncedNodes:  []*v1.Node{},
+		cloud:               cloud,
+		kubeClient:          kubeClient,
+		clusterName:         "test-cluster",
+		eventBroadcaster:    broadcaster,
+		eventRecorder:       recorder,
+		serviceLister:       serviceInformer.Lister(),
+		serviceListerSynced: serviceInformer.Informer().HasSynced,
+		nodeLister:          nodeInformer.Lister(),
+		nodeListerSynced:    nodeInformer.Informer().HasSynced,
+		serviceQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "service"),
+		nodeQueue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "node"),
+		lastSyncedNodes:     []*v1.Node{},
+	}
+
+	informerFactory.Start(stopCh)
+	informerFactory.WaitForCacheSync(stopCh)
```

Comment:
this is a noop if you don't pass the cache you want to wait to sync

https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L211

but for unit tests we always have to mock it 
```
controller.nodeListerSynced = alwaysReady
	controller.serviceListerSynced = alwaysReady
```

Reply:
Not sure I follow, this calls: https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/informers/factory.go#L141 

Which will just utilise `informer.HasSynced`, here: https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/informers/factory.go#L157

Reply:
It's not calling that. What you are linking to is the function `WaitForCacheSync`, the code is calling the pointer receiver `WaitForCacheSync` - which is what I linked to here: https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/informers/factory.go#L141

Reply:
doh, I should stop checking these things at night, sorry for the noise